Define bitfield type

primiti opened the pull request:
Hey @c-simmons  - you touched it last so you win!

I have a project that is going to store 100 of booleans in an aggregate.  This was going to take about 2K when marshaled because of the JSON overhead.  Instead, I added a bitfield type which allows them to be stored in a single string attribute.  This is almost a 20x space savings!

Please let me know if you have any questions.

ColinDKelley commented on lib/aggregate/attribute/bitfield.rb:
```diff
@@ -0,0 +1,26 @@
+
+  def self.available_options
+    super + [
+      :limit #
```
What is the # for?

primiti replied:
It was a comment that never happened.   Removed.

ColinDKelley commented on lib/aggregate/bitfield.rb:
```diff
@@ -0,0 +1,81 @@
+        false
+      when ' '
+        nil
+      when nil
```
You could write above
```
when ' ', nil
  nil
```

ColinDKelley commented on test/unit/bitfield_test.rb:
```diff
@@ -0,0 +1,85 @@
+    should "fill in details indexing past the length of the string" do
+      bitfield = Aggregate::Bitfield.new("")
+
+      bitfield[10] = true
```
Can you add an assertion? Like
```
assert_equal [nil]*10+[true], (0..10).map { |i| bitfield[i] }
```

primiti replied:
Yes, done.  Thanks

ColinDKelley commented on test/unit/bitfield_test.rb:
```diff
@@ -0,0 +1,85 @@
+          fail "Didn't raise like I expected"
+        rescue ArgumentError => ex
+          assert_equal "index out of bounds, index(5) >= limit(5)", ex.message
+        end
```
I think this can be written as
```
ex = assert_raises ArgumentError do
  bitfield[5]
end
assert_equal "index out of bounds, index(5) >= limit(5)", ex.message
```

primiti replied:
Thanks!

ColinDKelley commented on test/unit/combined_string_field_test.rb:
```diff
@@ -78,7 +78,7 @@ module ActiveRecordStub
         @instance.first = "abc\n123"
       rescue ArgumentError => ex
-        ex.message =~ /Cannot store newlines in combined fields storing \"abc\\n123\" in first/
+        assert ex.message =~ /Cannot store newlines in combined fields storing \"abc\\n123\" in first/
```
Again I think `assert_raises` works well. And the match can be explicitly tested:
```
assert_match /Cannot store newlines in combined fields storing \"abc\\n123\" in first/, ex.message
```
which is nicer because if it fails, we'll get a useful message instead of "false is not true".

primiti replied:
👍

c-simmons commented on lib/aggregate/bitfield.rb:
```diff
@@ -0,0 +1,81 @@
+    def []=(index, value)
+      check_index_limit(index)
+      if index > @string.size
+        @string = @string.ljust(index)
```
I would probably just skip the if since ljust(index) doesn't do anything when `index <= @string.size`

So just 
```
check_index_limit(index)
@string = @string.ljust(index)
@string[index] = to_character(value)
```
..

primiti replied:
thanks

primiti commented:
I have addressed all comments and the build is clean so I am merging with master.   (I am not worried about QA on this as the new sha's are not being used.
